### PR TITLE
Billion-scale vector search with Vespa - part two

### PR DESCRIPTION
**Pull request by jobergum**
Screenshot 2022-01-26 at 13 53 03

The summary needs more work - but thinking it's time to ask for feedback!

@geirst could you do a pass on the technical side, and @kkraune on messaging? 

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+reduce the storage and computational complexity of both exact and approximate nearest neighbor search. 
+This second post covers an experiment using a 1-billion binary-coded representation derived from a
+vector dataset used in the [big-ann-benchmark](http://big-ann-benchmarks.com/) challenge. The primary
+purpose of this post in the series is to highlight some of the trade-offs related to approximate nearest neighbor
+search and especially we focus on serving performance versus accuracy. 
+
+Vespa implements a version of the *HNSW (Hierarchical Navigable Small Word)*
+[algorithm](https://arxiv.org/abs/1603.09320) for approximate
+vector search. Before diving into this post, we recommend reading the [HNSW in Vespa](../approximate-nearest-neighbor-search-in-vespa-part-1/)
+blog post for why we choose the *HNSW* algorithm.
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:30]:
```suggestion
blog post explaining why we chose the *HNSW* algorithm.   
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+<img src="/assets/2022-01-27-billion-scale-knn-part-two/vincentiu-solomon-ln5drpv_ImI-unsplash.jpg"/>
+<p class="image-credit">
+Photo by <a href="https://unsplash.com/@vincentiu?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">
+Vincentiu Solomon</a> on <a href="https://unsplash.com/s/photos/stars?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
+</p>
+
+In the [first post](../billion-scale-knn) in this series, we introduced compact binary-coded vector representations that can
+reduce the storage and computational complexity of both exact and approximate nearest neighbor search. 
+This second post covers an experiment using a 1-billion binary-coded representation derived from a
+vector dataset used in the [big-ann-benchmark](http://big-ann-benchmarks.com/) challenge. The primary
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:23]:
```suggestion
vector dataset used in the [big-ann-benchmark](http://big-ann-benchmarks.com/) challenge. The 
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+purpose of this post in the series is to highlight some of the trade-offs related to approximate nearest neighbor
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:24]:
```suggestion
purpose of this post is to highlight some of the trade-offs related to approximate nearest neighbor
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+search and especially we focus on serving performance versus accuracy. 
+
+Vespa implements a version of the *HNSW (Hierarchical Navigable Small Word)*
+[algorithm](https://arxiv.org/abs/1603.09320) for approximate
+vector search. Before diving into this post, we recommend reading the [HNSW in Vespa](../approximate-nearest-neighbor-search-in-vespa-part-1/)
+blog post for why we choose the *HNSW* algorithm.   
+
+# Choosing a Vector Dataset
+When working with vector datasets and nearest neighbor search algorithms, using vectors from an
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:33]: I suggest replacing this whole paragraph by something like "When evaluating approximate nearest neighbor search algorithms it is important to use realistic vector data. If you don't have the vectors for your problem at hand, use a publicly available dataset."

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+For our experiments, we chose the [Microsoft SPACEV-1B](https://github.com/microsoft/SPTAG/tree/main/datasets/SPACEV1B)
+from Bing as our base vector dataset. 
+ >This is a dataset released by Microsoft from SpaceV, Bing web vector search scenario, for large
+scale vector search-related research usage. It consists of more than one billion document vectors
+and 29K+ query vectors encoded by the Microsoft SpaceV Superior model. 
+
+The vector dataset was published last year as part of the 
+[big-ann-benchmarks](http://big-ann-benchmarks.com/) challenge. 
+The vector dataset consists of one billion 100-dimensional vectors using `int8` precision. In other words,
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:46]:
```suggestion
It consists of one billion 100-dimensional vectors using `int8` precision. In other words,
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+* `int8` (8 bits, 1 byte) per dimension
+* `bfloat16` (16 bits, 2 bytes) per dimension
+* `float` (32 bits, 4 bytes) per dimension
+* `double` (64 bits, 8 bytes) per dimension
+
+Quantization and dimension reduction as part of the representation learning could save both
+memory and CPU cycles in the serving phase, and Microsoft researchers have undoublty had this in mind
+when using 100 dimensions with `int8` precision for the embedding. 
+
+Using the threshold function, we convert the mentioned *SPACEV-1B* vector dataset to a new and binarized
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:61]: Should you say something about why we do this?

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+<pre>
+&lt;nodes count=&quot;1&quot;&gt;
+  &lt;resources memory=&quot;256Gb&quot; vcpu=&quot;72&quot; disk=&quot;1000Gb&quot; disk-speed=&quot;fast&quot;/&gt;
+&lt;/nodes&gt;
+</pre>
+
+This deployment specification isolates resources used for feed and search, except for search and indexing related 
+resource usage on the content node. Isolating feed and search allows for easier on-demand resource scaling as the 
+stateless containers can be [auto-scaled](https://cloud.vespa.ai/en/autoscaling) faster with read and write
+volume than stateful content resources. However, for self-hosted deployments of Vespa, there is no node resource support or auto-scaling.
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:123]:
```suggestion
volume than stateful content resources. For self-hosted deployments of Vespa, you need to list the nodes you are using, and there is no auto-scaling.
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+      attribute {
+        distance-metric:hamming
+      }
+    }
+  }
+}
+</pre>
+<sub>*Vespa document schema without HNSW indexing enabled*</sub>
+
+# Evaluation Metrics
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:144]: This is a very good section!

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+a database of many fingerprints. Accepting any accuracy loss as measured by
+Recall@1 (*the* true closest neighbor) will have severe consequences for the overall usefulness of the
+application.  
+
+# Vector Indexing performance
+We want to quantify the impact of adding data structures for faster and approximate vector search on
+vector indexing throughput. We use the [Vespa HTTP feeding client](https://docs.vespa.ai/en/vespa-http-client.html) 
+to feed vector data to the Vespa instance. 
+
+It is expected that indexing performance is degraded when enabling *HNSW* indexing for
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:187]:
```suggestion
Writing becomes more expensive when enabling *HNSW* indexing for
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+approximate vector search. This is because insertion into the *HNSW* graph requires distance calculations and graph
+modifications which reduces overall throughput. Vespa's *HNSW* implementation uses multiple threads for
+distance calculations during indexing, but only a single writer thread can mutate the *HNSW* graph. 
+The single writer thread limits concurrency and resource utilization. Generally, Vespa balances CPU resources used for indexing versus searching
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:191]:
```suggestion
The single writer thread limits concurrency and resource consumption. Generally, Vespa balances CPU resources used for indexing versus searching
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+The real-time indexing throughput results are summarized in the following chart:
+<figure>
+    <img src="/assets/2022-01-27-billion-scale-knn-part-two/throughput.png" alt="Indexing performance"/>
+</figure>
+<sub>*Real-time indexing performance without HNSW indexing and with two HNSW parameter combinations.*</sub>
+
+Without *HNSW* enabled, Vespa is able to sustain 80 000 vector puts/s. By increasing the number of nodes in the 
+Vespa content cluster using Vespa's [content distribution](https://docs.vespa.ai/en/elastic-vespa.html), 
+it is possible to increase throughput horizontally. For example, using four nodes instead of one, would support 4x80 000 = 320 000 puts/.
+As we can see from the chart,  when we introduce *HNSW* indexing, the obtained real-time throughput drops significantly as it involves mutations of the
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:249]:
```suggestion
As we can see from the chart,  when we introduce *HNSW* indexing, the write throughput drops significantly as it involves mutations of the 
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+<figure>
+    <img src="/assets/2022-01-27-billion-scale-knn-part-two/throughput.png" alt="Indexing performance"/>
+</figure>
+<sub>*Real-time indexing performance without HNSW indexing and with two HNSW parameter combinations.*</sub>
+
+Without *HNSW* enabled, Vespa is able to sustain 80 000 vector puts/s. By increasing the number of nodes in the 
+Vespa content cluster using Vespa's [content distribution](https://docs.vespa.ai/en/elastic-vespa.html), 
+it is possible to increase throughput horizontally. For example, using four nodes instead of one, would support 4x80 000 = 320 000 puts/.
+As we can see from the chart,  when we introduce *HNSW* indexing, the obtained real-time throughput drops significantly as it involves mutations of the 
+*HNSW* graph and distance calculations.  In addition to indexing throughput, we also measure peak memory usage for the content node which is provided in the chart below:
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:250]:
```suggestion
*HNSW* graph and distance calculations.  In addition to indexing throughput, we also measure peak memory usage for the content node:
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+<figure>
+    <img src="/assets/2022-01-27-billion-scale-knn-part-two/memory.png" alt="Memory Usage(GB)"/>
+</figure>
+<sub>*Peak Memory Usage without HNSW indexing and with two HNSW parameter combinations.*</sub>
+
+Now, you might ask, why are Vespa using 64G of memory for this dataset in the baseline case without *HNSW*?  
+The reason is that Vespa stores the global documentid in memory, and the documentid consumes more memory than the vector
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:258]:
```suggestion
The reason is that Vespa stores the _global document id_ (a 96-bit hash of the document id string) in memory, and the document id consumes more memory than the vector
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+As we introduce *HNSW* indexing, the memory usage increases significantly due to the additional *HNSW* graph data structure which is also 
+in memory for fast access during searches and insertions. 
+
+# Brute-force exact nearest neighbor search performance 
+As we have seen in the indexing performance and memory utilization experiments, not using *HNSW* uses
+considerably less memory, and is the clear indexing throughput winner - but what about the search
+performance of brute force search? Without *HNSW* graph indexing, the complexity of the search for neighbors is linear with
+the total document volume, so that is surely slow for 1B documents?
+
+To overcome the latency issue, We can use one of the essential Vespa features: executing a query using multiple
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:272]:
```suggestion
To overcome the latency issue, We can use one of the essential Vespa features: Executing a query using multiple
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+[search threads](https://docs.vespa.ai/en/performance/sizing-search.html#num-threads-per-search). 
+By using more threads per query, Vespa can better use multi-CPU core architecture and
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:274]:
```suggestion
By using more threads per query, Vespa can make better use of multi-CPU core architectures and
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+reduce query latency at the cost of increased CPU resource usage per query. Most search libraries or
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:275]:
```suggestion
reduce query latency at the cost of increased CPU usage per query. Most search libraries or
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+engines require high concurrent query throughput to drive CPU utilization. 
+On the other hand, Vespa allows micro-slicing of the intra-node document volume so that multiple threads can execute the same
+query in parallel, each search thread working on a [partition](https://docs.vespa.ai/en/reference/schema-reference.html#num-search-partitions) 
+of the node's document volume. More threads per search lowers search latency, especially at the tail, at the cost of increased resource usage per query.
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:279]: A bit much repetition of the same point in this paragraph.

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+See more on using threads per search in the [Sizing and performance guide](https://docs.vespa.ai/en/performance/sizing-search.html#reduce-latency-with-multi-threaded-per-search-execution). 
+
+To easily test multiple threading configurations, we deploy multiple 
+Vespa [ranking profiles](https://docs.vespa.ai/en/ranking.html), chosing ranking profile is
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:283]:
```suggestion
Vespa [rank profiles](https://docs.vespa.ai/en/ranking.html). Choosing ranking profile is 
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+reduce query latency at the cost of increased CPU resource usage per query. Most search libraries or
+engines require high concurrent query throughput to drive CPU utilization. 
+On the other hand, Vespa allows micro-slicing of the intra-node document volume so that multiple threads can execute the same
+query in parallel, each search thread working on a [partition](https://docs.vespa.ai/en/reference/schema-reference.html#num-search-partitions) 
+of the node's document volume. More threads per search lowers search latency, especially at the tail, at the cost of increased resource usage per query.
+See more on using threads per search in the [Sizing and performance guide](https://docs.vespa.ai/en/performance/sizing-search.html#reduce-latency-with-multi-threaded-per-search-execution). 
+
+To easily test multiple threading configurations, we deploy multiple 
+Vespa [ranking profiles](https://docs.vespa.ai/en/ranking.html), chosing ranking profile is 
+a query time setting so it's easy to run experiments without having to re-deploy the application.
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:284]:
```suggestion
done in the query so it's easy to run experiments without having to re-deploy the application. 
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+* max-links-per-node = 8, neighbors-to-explore-at-insert 48 
+* max-links-per-node = 16, neighbors-to-explore-at-insert 96 
+
+We use the exact search to obtain the 100 ground truth exact nearest neighbors for all the queries in the dataset, and use those for the 
+approximate nearest neighbor search *recall@10* evaluation. We use 100 in the ground truth set to be able to take into account *distance ties*.
+
+With approximate nearest neighbor search in Vespa, the traversal of the HNSW graph uses one thread per
+search irrespective of the number of configured rank profile search threads; the threads are put to
+use if the ranking profile uses higher-level subsequent [ranking phases](https://docs.vespa.ai/en/phased-ranking.html).
+For use cases and ranking profiles without higher level ranking phases, it's recommended to explicit configure
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:327]:
```suggestion
For use cases and ranking profiles without higher level ranking phases, we recommend explicitly configuring
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+The *hnsw.exploreAdditionalHits* run-time parameter allows us to tune quality versus cost without re-building the *HNSW graph*.
+
+As we can see from the above graph, using more indexing resources with more links and more distance
+calculations during graph insertion improves the search quality at the same cost/latency. 
+As a result, we reach 90% recall@10 instead of 64% recall@10 at the exact cost of 4 ms search time. 
+Of course, the level of acceptable *recall@10* will be use case dependent, but the above figure illustrates the impact on search
+quality when using different construction HNSW parameters.
+
+Furthermore, comparing the 4ms at 90% recall@10 with the exact nearest neighbor search performance
+of 15000 ms, we achieve a speedup of 3,750x. Note that these are latency numbers for single-threaded searches.
```
Comment (bratseth) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:348]:
```suggestion
of 15000 ms, we see that we achieve a speedup of 3,750x. Note that these are latency numbers for single-threaded searches.
```

**Review by bratseth — APPROVED** [review]
This is great!

File: _posts/2022-01-27-billion-scale-knn-part-two.md (hunk 2 above)
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md:23]: Thanks, remove fluff

File: _posts/2022-01-27-billion-scale-knn-part-two.md (hunk 17 above)
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md:279]: Thanks, I've compressed it.

File: _posts/2022-01-27-billion-scale-knn-part-two.md (hunk 8 above)
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md:144]: Thanks!

File: _posts/2022-01-27-billion-scale-knn-part-two.md (hunk 6 above)
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md:61]: Done, point to first blog post.

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 22]
```diff
@@ -0,0 +1,369 @@
...
+As we introduce *HNSW* indexing, the memory usage increases significantly due to the additional *HNSW* graph data structure which is also 
+in memory for fast access during searches and insertions. 
+
+# Brute-force exact nearest neighbor search performance 
+As we have seen in the indexing performance and memory utilization experiments, not using *HNSW* uses
+considerably less memory, and is the clear indexing throughput winner - but what about the search
+performance of brute force search? Without *HNSW* graph indexing, the complexity of the search for neighbors is linear with
+the total document volume, so that is surely slow for 1B documents?
+
+To overcome the latency issue, We can use one of the essential Vespa features: Executing a query using multiple
```
Comment (kkraune) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:272]:
```suggestion
To overcome the latency issue, we can use one of the essential Vespa features: Executing a query using multiple
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+For example, with 4 ms average latency per search using one thread, a node with one CPU core will be able to evaluate up to about 250
+queries per second. 72 CPU cores would be 72x that, reaching 18,000 queries per second at 100% CPU
+utilization. Scaling further for increased query throughput is achieved using multiple replicas using grouped content
+distribution (Or more CPU cores per node). See more on how to size Vespa search deployments in the 
+[Vespa sizing guide](https://docs.vespa.ai/en/performance/sizing-search.html). 
+
+# Summary 
+In this blog post, we explored several trade-offs related to vector search.  
+We concluded that the quality, as measured by recall@k, must be weighed against the use case metrics 
+and the deployment cost.Furthermore, we demonstrated how multi-threaded search could reduce the
```
Comment (kkraune) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:358]:
```suggestion
and the deployment cost. Furthermore, we demonstrated how multi-threaded search could reduce the 
```

**Review by kkraune — APPROVED** [review]
Excellent post! This is a reference for people exploring ANN, pondering on cost/latency tradeoffs - will hence link as a further read from docs.vespa.ai guides. This is good for publishing - great work!

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 24]
```diff
@@ -0,0 +1,373 @@
...
+default value in Vespa is 16. The [HNSW paper](https://arxiv.org/abs/1603.09320) calls this parameter **M**.  
+A higher value of *max-links-per node* increases memory usage and reduces indexing throughput, but also improves the quality of the graph. 
+Note that the parameter also defines the memory consumption of the algorithm (which is proportional to *max-links-per-node*). 
+
+* **neighbors-to-explore-at-insert**  Specifies how many neighbors to explore when inserting a vector in
+the HNSW graph. The default value in Vespa is 200. This parameter is called **efConstruction** in the HNSW paper. 
+A higher value generally improves the quality but lowers indexing throughput as each insertion requires more
+distance computations. This parameter does not impact memory footprint. 
+
+We experiment with the following *HNSW* parameters combinations for evaluating feed indexing
```
Comment (geirst) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:208]:
```suggestion
We experiment with the following *HNSW* parameter combinations for evaluating feed indexing
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 25]
```diff
@@ -0,0 +1,373 @@
...
+ When we introduce *HNSW* indexing, the write-throughput drops significantly as it involves mutations of the
+*HNSW* graph and distance calculations.  In addition to indexing throughput, we also measure peak memory usage for the content node:
+
+<figure>
+    <img src="/assets/2022-01-27-billion-scale-knn-part-two/memory.png" alt="Memory Usage(GB)"/>
+</figure>
+<sub>*Peak Memory Usage without HNSW indexing and with two HNSW parameter combinations.*</sub>
+
+Now, you might ask, why are Vespa using 64G of memory for this dataset in the baseline case without *HNSW*?  
+The reason is that Vespa stores the _global document id_ (a 96-bit hash of the document id string) in memory, and the document id consumes more memory than the vector
```
Comment (geirst) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:265]: Consider adding a link to https://docs.vespa.ai/en/attributes.html#document-meta-store.

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 26]
```diff
@@ -0,0 +1,373 @@
...
+As we can see from the figure above, one search thread uses on average 15 seconds to compute the exact nearest neighbors. 
+By increasing the number of search threads per query to 32, we can reach sub-second search latency. 
+The catch is that at as low as one query per second (QPS), the node would be running at close to 100% CPU utilization. 
+Still, trading latency over throughput might be a good decision for some use cases that do not require high query throughput or 
+where CPU utilization is low (over-provisioned resources). In our case, 
+using all available CPU cores for our exact ground truth calculations reduced the overall time duration significantly. 
+
+# Approximate nearest neighbor search performance 
+Moving from exact to approximate nearest neighbor search, we evaluate the search performance
+versus accuracy using the recall@10 metric for the same *HNSW* combinations we used
```
Comment (geirst) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:319]:
```suggestion
versus accuracy using the recall@10 metric for the same *HNSW* parameter combinations we used
```

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 27]
```diff
@@ -0,0 +1,373 @@
...
+As a result, we reach 90% recall@10 instead of 64% recall@10 at the exact cost of 4 ms search time. 
+Of course, the level of acceptable *recall@10* will be use case dependent, but the above figure illustrates the impact on search
+quality when using different construction HNSW parameters.
+
+Furthermore, comparing the 4ms at 90% recall@10 with the exact nearest neighbor search performance
+of 15000 ms, we see that we achieve a speedup of 3,750x. Note that these are latency numbers for single-threaded searches.
+For example, with 4 ms average latency per search using one thread, a node with one CPU core will be able to evaluate up to about 250
+queries per second. 72 CPU cores would be 72x that, reaching 18,000 queries per second at 100% CPU
+utilization. Scaling further for increased query throughput is achieved using multiple replicas using grouped content
+distribution (Or more CPU cores per node). See more on how to size Vespa search deployments in the
```
Comment (geirst) [inline _posts/2022-01-27-billion-scale-knn-part-two.md:356]:
```suggestion
distribution (or more CPU cores per node). See more on how to size Vespa search deployments in the 
```

**Review by geirst — APPROVED** [review]
This is a great blog post! 👍

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+Quantization and dimension reduction as part of the representation learning could save both
+memory and CPU cycles in the serving phase, and Microsoft researchers have undoubtedly had this in mind
+when using 100 dimensions with `int8` precision for the embedding. 
+
+In the [first post](../billion-scale-knn)in this series we discussed some of the benefits of working with binarized vector datasets
+where the original vector representation (e.g, `int8`) is transformed to binary. 
+We don't train a binary representation model, but instead use the unsupervised *sign* threshold function. 
+Using the threshold function, we convert the mentioned *SPACEV-1B* vector dataset to a new and binarized
+dataset. Both queries and document vectors are binarized, and we use the [hamming distance
+metric](https://docs.vespa.ai/en/reference/schema-reference.html#distance-metric) for the NNS
```
[the remark this answers is not in the document]
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md]: Thanks!

File: _posts/2022-01-27-billion-scale-knn-part-two.md [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+13 dimensional dense single-order `int8` tensor in Vespa. 
+
+However, this transformation from *euclidean* to *hamming* does not preserve the
+original euclidean distance ordering, so we calculate a new set of ground truth nearest neighbors
+for the query dataset using *hamming* distance. Effectively, we create a new binarized vector dataset
+which we can use to experiment and demonstrate some trade-offs related to vector search:
+
+* Brute force nearest neighbor search using multiple search threads to parallelize the search for
+exact nearest neighbors 
+* Approximate nearest neighbor search using where we accept an accuracy loss compared
```
[the remark this answers is not in the document]
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md]: Fixed

File: _posts/2022-01-27-billion-scale-knn-part-two.md (hunk 25 above)
Comment (jobergum) [reply _posts/2022-01-27-billion-scale-knn-part-two.md:264]: Done

**Pull request merged**